### PR TITLE
AP_Hygrometer: add hygrometer sensor 

### PR DESCRIPTION
**Pull request by gitfishup**
The flight controller receive the hygrometer message from CAN device and flight controller send hygrometer message to the ground station.

background:
CUAV developed a new airspeed sensor that can be heated.The airspeed sensor heats up according to the set temperature, so as to prevent airspeed failure due to ice on the head of the airspeed sensor when the fixed-wing UAV is used in high and cold places.The temperature and humidity sensor is used to detect the temperature and humidity of the pitot tube head. If the measured temperature is lower than the set temperature, the airspeed sensor will enable heating to prevent ice.

The value shows in MissionPlanner.
lQLPDhrQcGhJCznNAhzNAaaw9Ik0twoGiU8Bgr_Y7EAcAA_422_540

Logging:
[hygrometer_log.zip](https://github.com/ArduPilot/ardupilot/files/7451226/hygrometer_log.zip)
 
parameters for setting:
HYGROx_TYPE:  0 for none , 1 for uavcan. Note: HYGRO2_TYPE can be set after HYGRO1_TYPE is set to 1;
HYGRO_PRIMARY: to select the primary one.
image

### CODE REVIEW

**Comment by gitfishup** [comment]
@peterbarker Could you please have a review?

File: libraries/AP_Hygrometer/AP_Hygrometer.cpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    for (uint8_t i=0; i<HYGROMETER_MAX_SENSORS; i++) {
+         switch ((enum Type)param[i].type.get()) {
+             case Type::NONE:
+                // nothing to do
+                break;
+             case Type::UAVCAN:
+#if HAL_ENABLE_LIBUAVCAN_DRIVERS    
+    sensor[i] = AP_Hygrometer_UAVCAN::probe(*this, i, state[i], param[i]);
+#endif
+                break;
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.cpp:76]:
```suggestion
#if HAL_ENABLE_LIBUAVCAN_DRIVERS    
             case Type::UAVCAN:
                sensor[i] = AP_Hygrometer_UAVCAN::probe(*this, i, state[i], param[i]);
                break;
#endif
```

File: libraries/AP_Hygrometer/AP_Hygrometer.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    bool get_humidity(float &humidity) { return get_humidity(primary, humidity); }
+
+    bool get_id(uint8_t i, uint8_t &id); 
+    bool get_id(uint8_t &id) { return get_id(primary, id); }
+
+    static const struct AP_Param::GroupInfo var_info[];
+
+    enum class Type {
+        NONE=0,
+        UAVCAN=1,
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.h:45]:
```suggestion
#if HAL_ENABLE_LIBUAVCAN_DRIVERS
        UAVCAN=1,
#endif
```

This makes it harder to leave code in accidentally :-)

File: libraries/AP_Hygrometer/AP_Hygrometer.cpp [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+         }
+        if (sensor[i] && !sensor[i]->init()) {
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.cpp:78]:
```suggestion
        if (sensor[i] == nullptr) {
            continue;
        }
        if (!sensor[i]->init()) {
```

File: libraries/AP_Hygrometer/AP_Hygrometer.cpp [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Hygrometer %u init failed", i + 1);
+            delete sensor[i];
+            sensor[i] = nullptr;
+        }
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.cpp:82]:
```suggestion
            continue;
        }
```

File: libraries/AP_Hygrometer/AP_Hygrometer.cpp [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        if (sensor[i] != nullptr) {
+            num_sensors = i+1;
+        }
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.cpp:85]:
```suggestion
        num_sensors = i+1;
```

File: libraries/AP_Hygrometer/AP_Hygrometer.cpp [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+bool AP_Hygrometer::get_humidity(uint8_t i, float &humidity)
+{
+    if (!enabled(i)) {
+        return false;
+    }
+    if (sensor[i]) {
+        return sensor[i]->get_humidity(humidity);
+    }
+
+    return false;
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.cpp:108]:
```suggestion
    if (sensor[i] == nullptr) {
        return false;
    }

    return sensor[i]->get_humidity(humidity);
```

File: libraries/AP_Hygrometer/AP_Hygrometer.cpp [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+{
+    if (!enabled(i)) {
+        return false;
+
+    }
+    if (sensor[i]) {
+        return sensor[i]->get_temperature(temperature);
+    }
+
+    return false;
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.cpp:121]:
```suggestion
    if (sensor[i] == nullptr) {
        return false;
    }

    return sensor[i]->get_temperature(temperature);
```

File: libraries/AP_Hygrometer/AP_Hygrometer.cpp [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+{
+    if (!enabled(i)) {
+        return false;
+    }
+    if (sensor[i]) {
+        return sensor[i]->get_id(id);
+    }
+    return false;
+}
+
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.cpp:134]:
```suggestion
```

Unused

File: libraries/AP_Hygrometer/AP_Hygrometer.cpp [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+inline void AP_Hygrometer::Log_HYGR()
+{
+    const uint64_t now = AP_HAL::micros64();
+    for (uint8_t i=0; i<HYGROMETER_MAX_SENSORS; i++) {
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.cpp:138]:
```suggestion
    for (uint8_t i=0; i<num_sensors; i++) {
```

File: libraries/AP_Hygrometer/AP_Hygrometer.cpp [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        if (!enabled(i)) {
+            continue;
+        }
+        float temperature,humidity;
+        if (!get_temperature(i, temperature)) {
+            temperature = 0;
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.cpp:144]:
```suggestion
            temperature = AP::logger().quiet_nanf();
```

File: libraries/AP_Hygrometer/AP_Hygrometer.cpp [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+        }
+        if (!get_humidity(i, humidity)) {
+            humidity = 0;
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.cpp:147]:
```suggestion
            humidity = AP::logger().quiet_nanf();
```

File: libraries/AP_Hygrometer/AP_Hygrometer.cpp [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+        }
+
+        state[i].temperature = temperature;
+        state[i].humidity = humidity;
+
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.cpp:152]:
```suggestion
```

Logging function should not be modifying state.  If you want to grab data from the backends, do it as another function call out of your `::update` function - and only use (e.g.) `state[i].temperature` in here.

File: libraries/AP_Hygrometer/AP_Hygrometer.cpp [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+namespace AP {
+
+AP_Hygrometer *hygrometer()
+{
+    return AP_Hygrometer::get_singleton();
+}
+
+};
+
+#endif // HAL_HYGROMETER_ENABLED
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.cpp:200]:
```suggestion
#endif // HAL_HYGROMETER_ENABLED
```

Comment (peterbarker) [reply libraries/AP_Hygrometer/AP_Hygrometer.cpp:200]: Please ensure all files have a lf at the end

File: libraries/AP_Hygrometer/AP_Hygrometer.h [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+#if HAL_HYGROMETER_ENABLED
+
+#include <stdint.h>
+#include "stdio.h"
+#include <AP_Param/AP_Param.h>
+#include <AP_Logger/AP_Logger.h>
+
+#ifndef HYGROMETER_MAX_SENSORS
+#define HYGROMETER_MAX_SENSORS  2
+#endif
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.h:18]:
```suggestion
#ifndef AP_HYGROMETER_MAX_SENSORS
#define AP_HYGROMETER_MAX_SENSORS  2
#endif
```

File: libraries/AP_Hygrometer/AP_Hygrometer.h [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+    bool get_temperature(uint8_t i, float &temperature) WARN_IF_UNUSED;
+    bool get_temperature(float &temperature) { return get_temperature(primary, temperature); }
+
+    bool get_humidity(uint8_t i, float &humidity);
+    bool get_humidity(float &humidity) { return get_humidity(primary, humidity); }
+
+    bool get_id(uint8_t i, uint8_t &id); 
+    bool get_id(uint8_t &id) { return get_id(primary, id); }
+
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.h:40]:
```suggestion
```

File: libraries/AP_Hygrometer/AP_Hygrometer.h [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    struct State {
+        float temperature;
+        float humidity;
+    };
+
+    struct Param {
+         AP_Enum<Type> type;
+    };
+
+        // return true if hygrometer is enabled
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.h:57]:
```suggestion
    // return true if hygrometer is enabled
```

File: libraries/AP_Hygrometer/AP_Hygrometer.h [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    bool enabled(uint8_t i) const {
+        if (i < HYGROMETER_MAX_SENSORS) {
+             return (Type)param[i].type.get() != Type::NONE;
+        }
+        return false;
+    }
+    bool enabled(void) const { return enabled(primary); }
+
+    // get current primary sensor
+    uint8_t get_primary(void) const { return primary; }
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.h:67]: We probably don't need an accessor for this.

File: libraries/AP_Hygrometer/AP_Hygrometer.h [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+    static AP_Hygrometer *_singleton;
+
+    AP_Int8 primary_sensor;
+
+    Param param[HYGROMETER_MAX_SENSORS];
+
+    State state[HYGROMETER_MAX_SENSORS];
+
+    // current primary sensor
+    uint8_t primary;
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer.h:87]: You've got this and `primary_sensor` - pretty sure that's not right.

File: libraries/AP_Hygrometer/AP_Hygrometer_UAVCAN.cpp [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+#define C_TO_KELVIN 273.15f
+
+UC_REGISTRY_BINDER(HygrometerCb, dronecan::sensors::hygrometer::Hygrometer);
+
+AP_Hygrometer_UAVCAN::DetectedModules AP_Hygrometer_UAVCAN::_detected_modules[] = {0};
+HAL_Semaphore AP_Hygrometer_UAVCAN::_sem_registry;
+
+AP_Hygrometer_UAVCAN::AP_Hygrometer_UAVCAN(AP_Hygrometer &_frontend, uint8_t _instance, AP_Hygrometer::State &_state, AP_Hygrometer::Param &_params) :
+    AP_Hygrometer_Backend(_frontend, _instance, _state, _params)
+{}
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer_UAVCAN.cpp:36]: We probably do not need this.  See comment in header file

File: libraries/AP_Hygrometer/AP_Hygrometer_Backend.h [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+
+class AP_Hygrometer_Backend
+{
+public:
+    AP_Hygrometer_Backend(AP_Hygrometer &_frontend, uint8_t _instance, AP_Hygrometer::State &_state, AP_Hygrometer::Param &_params);
+    
+    virtual ~AP_Hygrometer_Backend(void) { }
+   
+    float get_humidity() const { return state.humidity; }
+    float get_temperature() const { return state.temperature; }
```
Comment (peterbarker) [inline libraries/AP_Hygrometer/AP_Hygrometer_Backend.h:16]:
```suggestion
    virtual float get_humidity() const = 0;
    virtual float get_temperature() const = 0;
```

**Review by peterbarker — CHANGES_REQUESTED** [review]
This is a partial review only.

I think you need to comment in here somewhere how the data moves around, and what the protections are.

Often what you will have is your `::update()` function (which runs on the main thread) will call each `backend[i].update()` which takes a sempahore and copies data from backend-private data into `state[i]`.  You're kind of doing that here, but mixing it with the logging function :-)  Please split those apart.

I haven't looked through the UAVCAN driver yet.

**Pull request closed without merging**